doc/man/rgbcorrelator.1
.TH rgbcorrelator 1 "March 2012" "ESRF" "PyMca X-Ray Fluorescence Toolkit"

.SH NAME

rgbcorrelator - PyMca image correlation analysis application

.SH SYNOPSIS

rgbcorrelator [FILE1] [FILE2] ...

.SH DESCRIPTION

.P
This an alias to pymcapostbatch

Start the graphical user interface of the PyMca X-Ray 
Fluorescence Toolkit rgbcorrelator tool.

It allows one to superpose up to three images in RED, GREEN and BLUE in 
order to investigate correlations among them. All the images have to 
be of the same dimension.

The program builds a table with all the input images but only three 
of them can be superposed at the same time.

This program also offers the possibility to perform mathematical operations 
among the images and simple multivariate analysis.


.SH EXAMPLES

.B rgbcorrelator
.P
Start the program with a file browser to select the input files.

.B rgbcorrelator file1.tif file2.tif file3.tif file4.tif
.P
Tries to open the files named file1.tif, file2.tif, file3.tif and 
file4.tif.

.SH SEE ALSO
ImageJ


changelog.txt
VERSION 4.7.4

Support readout of top level datasets in HDF5 files.

Improved handling of Tracer csv files.

Correct problems reading spec 6.02.07+ shared memory.

VERSION 4.7.3

Correct bug using SNIP background with the Fast XRF linear fit stack plugin.

Better handling of screen resolutions with 768 vertical pixels.

VERSION 4.7.2

Allow fitting of all Cr L lines

Correct bug saving image alignment data to disk.

Correct bug using linear polynomial background with the Fast XRF linear fit stack plugin

VERSION 4.7.1

Support SOLEIL multiple-file different-scan-number maps.

Support batch fitting of multiple HDF5 files.

Fast XRF linear fit stack plugin.

Possibility to load image shifts in the image alignment stack plugin.

Basic support of MRC file format.

Add ID08 advanced alignment scan plugin.

VERSION 4.7.0

Add basic support for calculating multiple excitation corrections via the
XMI-MSIM Monte Carlo code.

Add Image alignment capabilities to the ROI Imaging tool.

Improved handling of HDF5 external links.

Solve printing issue when printing from the File menu while the SCAN window is
active

Recover reading support of SPE files from SLAC

Correct reading of .chi files maps

Extend the maximum number of counters supported in a specfile.

Prevent endless loop in specfile when the last character limiting a scan is '#'

Correct bug plotting a set of MCAs as one image when the number of MCA
detectors is more than one.

Background subtraction methods also work on stack browsers.

Handling of scales in image plots.

Add ID26 RIXS plugin as part of MultiScanToMesh plugins.

Add XMCD and XAS plugins from ID08 and ID12.

Add plugin to align curves.

Stacks of DESY fio files readable by windows binary.

Reduce memory usage when working with large in-memory stacks.

Decide to load dynamically the data based on the amount of physical memory
installed instead of based on a fixed data size.

VERSION 4.6.2

Fix calculation of the MCA associated to a region when using dynamically
loaded 1D stacks.

Fix scan XANES normalization. All curves normalized instead of just the active one.

VERSION 4.6.1

Support GZIP compressed EDF files (extensions: edf.gz, ccd.gz, raw.gz)

Fix annoying issue of having to select the left side of a pixel in order to
have the pixel actually selected. 

Fix windows problem of PyMca not starting when the "My Documents" folder has
been relocated (issue number 3537267)

Fix Debian hurd-i386 build from source problem.


VERSION 4.6.0

Add an X-ray fluorescence spectrum for training purposes.

Improved unicode support.

Added a simple stack normalization plugin to the ROI tool.

Added a simple XANES normalization plugins to the ROI tool and the scan
window.

Adapt installation script to simplify linux distribution maintainers task.

Add a minimalistic set of post-installation tests.

Add man pages.

Correct variance information displayed in the terminal window.

Restore functionality of colormap-based mask calculation.

Add a progress bar to the stack simple fitting plugin.

Possibility to perform simple fitting of masked stack regions. 

VERSION 4.5.0

Support h5py version 2.x

Support OMNIC 8.x .map file format

Support TIFF format. Uncompressed and packbits.

Support a couple of ASCII based file formats of beamlines of Diamond and APS.

Preliminary Python 3.2 support.

Support principal components analysis of dynamically loaded stacks.

Improved HDF5 file format handling.

Possibility to visualize variable width line profiles on images.

Add split Gauss, split Lorentz and split Pseudo-Voigt as fitting functions.

Do not automatically sort the list of files to be treated in batch mode. The
program will respect the order provided by the user. 

Allow visualized 3D objects to share the same colormap.

Lower the energy threshold in order to consider Boron K X rays.

Correct Arsenic density.

Allow to customize the saving of Scan window curves using matplotlib.

Allow multiple processes in MacOS X when using a non frozen PyMca version.

VERSION 4.4.1

Allow detailed customization of 1D output.

Add simple normalization plugins to the Scan window.

Implement a plugin system for the ROI Imaging tool.

Implement a generic batch fitting of stacks as a stack plugin. The functions are totally specified by the user.

Implement simple background removal tools as stack plugins.

Implement an alternative ROI window as a stack plugin. 

Support convertion of big EDF image stacks to HDF5.

Visualization of large image stacks via dynamic loading and use of HDF5.

Support MDP based PCA and ICA on user selected regions.

Read calibration from OMNIC 7.x .map files.

Support Bruker Opus maps exported in DPT format.

Improve the reading speed of TwinMic .dta files.

Add simple MarCCD support.

Add very simple support for some flavours of Pilatus CBF.

Show file header information.

Properly handle Amptek MCA calibration when more than two points or ROIs have been used.

Allow Savitzky-Golay (SG) filtering of stacks.

Correct problem affecting odd order SG derivatives.

Very primitive HDF5 2D Visualization.

Possibility to visualize HDF5 3D datasets as series of images.

Implement a context menu on the main window HDF5 browser.

Table visualization of non-numerical HDF5 datasets.

Support segmented HDF5 files using default segmentation.

llow the use of several processes when batch fitting a single HDF5 file.

Allow reading of pure image formats in the RGB correlator.

The fit configuration window was too high for Mac computers having exactly 800
pixels vertical resolution.

Workaround a windows problem when trying to select thousands of files in the
batch file dialog by using Qt file dialogs.

Minor visualization and data handling improvements.

Prevent a crash when the excitation energy is below 1 keV.

Add a script to build PyMca from source on Debian or Ubuntu systems.

Make specfile LC_NUMERIC locale independent to solve Debian bug report 602471.


PyMca5/PyMcaGui/RGBCorrelator.py
#!/usr/bin/env python
#/*##########################################################################
# Copyright (C) 2004-2014 V.A. Sole, European Synchrotron Radiation Facility
#
# This file is part of the PyMca X-ray Fluorescence Toolkit developed at
# the ESRF by the Software group.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This file is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Please contact the ESRF industrial unit (industry@esrf.fr) if this license
# is a problem for you.
#
#############################################################################*/
__author__ = "V.A. Sole - ESRF Data Analysis"
__contact__ = "sole@esrf.fr"
__license__ = "LGPL2+"
__copyright__ = "European Synchrotron Radiation Facility, Grenoble, France"
import sys
import os
import numpy
from PyMca5.PyMcaGui.pymca import RGBCorrelatorWidget
qt = RGBCorrelatorWidget.qt
if hasattr(qt, 'QString'):
    QString = qt.QString
else:
    QString = str
from PyMca5.PyMcaGui import RGBCorrelatorGraph
from PyMca5.PyMcaGui import QPyMcaMatplotlibSave
MATPLOTLIB = True


class RGBCorrelator(qt.QWidget):
    sigRGBCorrelatorSignal = qt.pyqtSignal(object)
    def __init__(self, parent = None, graph = None, bgrx = True):
        qt.QWidget.__init__(self, parent)
        self.setWindowTitle("PyMca RGB Correlator")
        self.setWindowIcon(qt.QIcon(qt.QPixmap(RGBCorrelatorGraph.IconDict['gioconda16'])))
        self.mainLayout = qt.QVBoxLayout(self)
        self.mainLayout.setContentsMargins(0, 0, 0, 0)
        self.mainLayout.setSpacing(6)
        self.splitter   = qt.QSplitter(self)
        self.splitter.setOrientation(qt.Qt.Horizontal)
        self.controller = RGBCorrelatorWidget.RGBCorrelatorWidget(self.splitter)
        self._y1AxisInverted = False
        self._imageBuffer = None
        self._matplotlibSaveImage = None
        standaloneSaving = True
        if graph is None:
            if MATPLOTLIB:
                standaloneSaving = False
            self.graphWidget = RGBCorrelatorGraph.RGBCorrelatorGraph(self.splitter,
                                            standalonesave=standaloneSaving)
            if not standaloneSaving:
                self.graphWidget.saveToolButton.clicked[()].connect( \
                         self._saveToolButtonSignal)
                self._saveMenu = qt.QMenu()
                self._saveMenu.addAction(QString("Standard"),    self.graphWidget._saveIconSignal)
                self._saveMenu.addAction(QString("Matplotlib") , self._saveMatplotlibImage)
            self.graph = self.graphWidget.graph
            #add flip Icon
            self.graphWidget.hFlipToolButton.clicked[()].connect( \
                         self._hFlipIconSignal)
            self._handleGraph    = True
        else:
            self.graph = graph
            self._handleGraph = False 
        #self.splitter.setStretchFactor(0,1)
        #self.splitter.setStretchFactor(1,1)
        self.mainLayout.addWidget(self.splitter)
        
        self.reset    = self.controller.reset
        self.addImage = self.controller.addImage
        self.removeImage = self.controller.removeImage
        self.addImageSlot = self.controller.addImageSlot
        self.removeImageSlot = self.controller.removeImageSlot
        self.replaceImageSlot = self.controller.replaceImageSlot
        self.setImageShape = self.controller.setImageShape
        self.update   = self.controller.update
        self.transposeImages   = self.controller.transposeImages
        self.controller.sigRGBCorrelatorWidgetSignal.connect( \
                     self.correlatorSignalSlot)

    def _hFlipIconSignal(self):
        if self._handleGraph:
            if self.graph.isYAxisInverted():
                self.graph.invertYAxis(False)
            else:
                self.graph.invertYAxis(True)
            self.graph.replot()
            #this is not needed
            #self.controller.update()
            return

    def correlatorSignalSlot(self, ddict):
        if 'image' in ddict:
            # keep the image buffer as an array
            self._imageBuffer = ddict['image']
            size = ddict['size']
            self._imageBuffer.shape = size[1],size[0],4
            self._imageBuffer[:,:,3] = 255
            self.graph.addImage(self._imageBuffer)
            self.graph.replot()

    def _saveToolButtonSignal(self):
        self._saveMenu.exec_(self.cursor().pos())

    def _saveMatplotlibImage(self):
        if self._matplotlibSaveImage is None:
            self._matplotlibSaveImage = QPyMcaMatplotlibSave.SaveImageSetup(None,
                                                                            None)
            self._matplotlibSaveImage.setWindowTitle("Matplotlib RGBCorrelator")

        #Qt is BGR while the others are RGB ...
        self._matplotlibSaveImage.setPixmapImage(self._imageBuffer, bgr=True)
        self._matplotlibSaveImage.show()
        self._matplotlibSaveImage.raise_()


    def closeEvent(self, event):
        ddict = {}
        ddict['event'] = "RGBCorrelatorClosed"
        ddict['id']    = id(self)
        self.controller.close()
        if self._matplotlibSaveImage is not None:
            self._matplotlibSaveImage.close()
        self.sigRGBCorrelatorSignal.emit(ddict)
        qt.QWidget.closeEvent(self, event)

    def show(self):
        if self.controller.isHidden():
            self.controller.show()
        qt.QWidget.show(self)

def test():
    app = qt.QApplication([])
    app.lastWindowClosed.connect(app.quit)
    if 0:
        graphWidget = RGBCorrelatorGraph.RGBCorrelatorGraph()
        graph = graphWidget.graph
        w = RGBCorrelator(graph=graph)
    else:
        w = RGBCorrelator()
        w.resize(800, 600)
    import getopt
    options=''
    longoptions=[]
    opts, args = getopt.getopt(
                    sys.argv[1:],
                    options,
                    longoptions)      
    for opt,arg in opts:
        pass
    filelist=args
    if len(filelist):
        try:
            import DataSource
            DataReader = DataSource.DataSource
        except:
            import EdfFileDataSource
            DataReader = EdfFileDataSource.EdfFileDataSource
        for fname in filelist:
            source = DataReader(fname)
            for key in source.getSourceInfo()['KeyList']:
                dataObject = source.getDataObject(key)
                w.addImage(dataObject.data, os.path.basename(fname)+" "+key)
    else:
        print("This is a just test method using 100 x 100 matrices.")
        print("Run PyMcaPostBatch to have file loading capabilities.") 
        array1 = numpy.arange(10000)
        array2 = numpy.resize(numpy.arange(10000), (100, 100))
        array2 = numpy.transpose(array2)
        array3 = array1 * 1
        w.addImage(array1)
        w.addImage(array2)
        w.addImage(array3)
        w.setImageShape([100, 100])
    w.show()
    app.exec_()

if __name__ == "__main__":
    test()
        
